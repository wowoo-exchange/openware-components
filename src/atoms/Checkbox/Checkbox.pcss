.cr-checkbox {
  cursor: pointer;
  display: block;
  margin-bottom: calc(1.2 * var(--checkbox-general-intent));
  padding-left: calc(3.6 * var(--checkbox-general-intent));
  position: relative;
  user-select: none;

  &__checkitem {
    position: absolute;
    top: 0;
    left: 0;
    height: calc(2.4 * var(--checkbox-size));
    width: calc(2.4 * var(--checkbox-size));
    background-color: var(--checkbox-background);
    border: 1px solid var(--checkbox-border-color);

    &::after {
      content: "";
      display: none;
      margin-left: calc(0.7 * var(--checkbox-general-intent));
      width: calc(0.8 * var(--checkbox-symbol-size));
      height: calc(1.6 * var(--checkbox-symbol-size));
      border: solid var(--checkbox-symbol-color);
      border-width: 0 2px 2px 0;
      transform: rotate(45deg);
    }
  }

  &__label {
    display: inline-block;
    font-size: var(--checkbox-label-font-size);
    padding: calc(0.6 * var(--checkbox-general-intent));
    padding-left: var(--checkbox-general-intent);
    vertical-align: top;
    white-space: nowrap;
  }

  &:hover {
    input ~ &__checkitem {
      background-color: var(--checkbox-hover-background);
    }
  }

  &__disabled {
    .cr-checkbox__label,
    .cr-checkbox__checkitem {
      opacity: 0.7;
    }
  }

  input {
    cursor: pointer;
    opacity: 0;
    position: absolute;

    &:checked ~ .cr-checkbox__checkitem {
      background-color: var(--checkbox-checked-background);

      &::after {
        display: block;
      }
    }
  }
}
