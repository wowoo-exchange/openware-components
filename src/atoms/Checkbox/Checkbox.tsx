import * as React from 'react';

import classnames from 'classnames';

interface CheckboxProps {
  /**
   * Property for setting or getting status of Checkbox component
   */
  checked?: boolean;
  /**
   * Additional class name. By default element receives `cr-checkbox` class
   * @default empty
   */
  className?: string;
  /**
   * Property for disabling Checkbox component
   */
  disabled?: boolean;
  /**
   * Property for label of Checkbox component
   */
  label: string;
  /**
   * Function for getting event of changing status of checkbox
   */
  onChange?: () => void;
  /**
   * Function for getting event of changing status of checkbox
   */
  slider?: boolean;
}

const Checkbox: React.FunctionComponent<CheckboxProps> = props => {
  const {
    checked,
    className,
    disabled,
    label,
    onChange,
    slider,
  } = props;
  const cx = classnames('cr-checkbox', {
    'cr-checkbox__disabled': disabled,
  }, className);
  return (
    <label className={cx}>
      <input
        checked={checked}
        className="cr-checkbox__input"
        onChange={onChange}
        type="checkbox"
      />
      <span className={slider ? 'slider' : 'cr-checkbox__checkitem'} />
      <span className="cr-checkbox__label">{label}</span>
    </label>
  );
};

export {
  Checkbox,
  CheckboxProps,
};
