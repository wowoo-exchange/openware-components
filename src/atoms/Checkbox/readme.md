Example of usage:

```js
const handleCheckboxChange = (e) => {
  console.log(e.target.checked);
}

<div>
  <Checkbox
    label="Default checkbox"
    onChange={handleCheckboxChange}
  />
  <Checkbox
    checked={true}
    onChange={handleCheckboxChange}
    disabled
    label="Disabled checked checkbox"
  />
  <Checkbox
    checked={false}
    onChange={handleCheckboxChange}
    disabled
    label="Disabled unchecked checkbox"
  />
</div>
```
