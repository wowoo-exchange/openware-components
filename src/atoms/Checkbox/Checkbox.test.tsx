import {shallow, ShallowWrapper} from 'enzyme';
import * as React from 'react';
import { Checkbox, CheckboxProps } from './Checkbox';

const defaultProps: CheckboxProps = {
  disabled: false,
  label: 'text',
};

const setup = (props: Partial<CheckboxProps> = {}) =>
  shallow(<Checkbox {...{ ...defaultProps, ...props }} />);

describe('Checkbox', () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = setup();
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        expect(wrapper.hasClass('cr-checkbox')).toBeTruthy();
    });

    it('should pass along supplied className', () => {
        const className = 'some-class';
        wrapper = setup({ className });
        expect(wrapper.hasClass(className)).toBeTruthy();
    });

    it('should have correct className when disabled is true', () => {
        wrapper = setup({ disabled: true });
        expect(wrapper.hasClass('cr-checkbox__disabled')).toBeTruthy();
    });

    it('should have correct className when slider is true', () => {
        wrapper = setup({ slider: true });
        expect(wrapper.find('span').first().hasClass('slider')).toBeTruthy();
    });

    it('checkitem element should have correct class', () => {
        const { className } = wrapper.find('span').first().props();
        expect(className).toContain('cr-checkbox__checkitem');
    });

    it('checkitem element should have correct class', () => {
        const labelProps = wrapper.find('span').last().props();
        const { children } = labelProps;
        const { className } = labelProps;
        expect(children).toContain('text');
        expect(className).toContain('cr-checkbox__label');
    });
});
