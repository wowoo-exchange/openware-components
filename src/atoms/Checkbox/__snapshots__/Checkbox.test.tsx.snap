// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render 1`] = `
<label
  className="cr-checkbox"
>
  <input
    className="cr-checkbox__input"
    type="checkbox"
  />
  <span
    className="cr-checkbox__checkitem"
  />
  <span
    className="cr-checkbox__label"
  >
    text
  </span>
</label>
`;
